tsinfer/formats.py

```diff
@@ -2480,6 +2480,127 @@ def insert_proxy_samples(
         assert other.num_ancestors == self.num_ancestors + len(sample_ids)
         return other
 
+    def truncate_ancestors(
+        self,
+        lower_time_bound,
+        upper_time_bound,
+        length_multiplier=2,
+        **kwargs,
+    ):
+        """
+        Truncates the length of ancestors above a given time and returns a new
+        :class:`.AncestorData` instance.
+
+        The maximum length of an ancestor existing between ``lower_time_bound'' and
+        ``upper_time_bound'', multiplied by ``length_multplier'', is used to truncate
+        the length of all ancestors older than ``upper_time_bound'' which contain
+        at least one ``focal_site''.
+
+        Specifically, the haplotypes of ancestors older than ``upper_time_bound'' may
+        extend no further than *half of this maximum length* to the either side of the
+        leftmost and rightmost focal sites on each ancestor. Note that ancestors above
+        ``upper_time_bound'' may still be longer than the maximum length if the
+        ancestor contains greater than 2 ``focal_sites''.
+
+        This function should be used when :func:`tsinfer.generate_ancestors` generates
+        old ancestors which are very long, as this can adversely affect performance.
+        Older ancestors should generally be shorter than younger ancestors, so
+        truncating the lengths of older ancestors has negligible effect on inference
+        accuracy.
+
+        :param float lower_time_bound: Defines the lower bound of the closed interval
```

Comment:
Usually in tskit we make intervals open-closed, so it might be confusing to have a fully closed interval here. It'll make no difference in practise. Might be worth writing down the operation in psuedocode

```
Given a set of haplotypes H such that lower_time_bound <= h.time < upper_time_bound, we let max_len = max(max(h.length) for h in H). Then, we truncate all haplotypes where h.time >= upper.
```